[libSyntax] Several fixes for libSyntax parsing

ahoppen opened the pull request:
Fixe a couple of bugs in libSyntax parsing found by enabling `-verify-syntax-tree` for `%target-build-swift`:
- Fix parsing of the `actor` contextual keyword in actor decls
- Don't build a libSyntax tree when parsing the availability macro
  - The availability macro is not part of the source code and doesn't form a valid Swift file, thus creation of a libSyntax tree is completely pointless and will fail
- Add support for parsing `@_originallyDefinedIn` attributes.
- Add support for parsing `#sourceLocation` in member decl lists
- Add support for effectful properties (throwing/async getters/setters)
- Add support for optional types as the base of a key path (e.g. `\TestOptional2?.something`)
- Allow platform restrictions without a version (e.g. `_iOS13Aligned`)

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
https://github.com/apple/swift-syntax/pull/274

@swift-ci Please test macOS

rintaro commented on utils/gyb_syntax_support/DeclNodes.py:
```diff
@@ -532,6 +532,15 @@
                       'mutableAddressWithNativeOwner',
                       '_read', '_modify'
                    ]),
+             Child('AsyncKeyword', kind='IdentifierToken',
+                   classification='Keyword',
+                   text_choices=['async'], is_optional=True),
+             Child('ThrowsKeyword', kind='Token',
+                   is_optional=True,
+                   token_choices=[
+                       'ThrowsToken',
+                       'RethrowsToken',
+                   ]),
              Child('Parameter', kind='AccessorParameter', is_optional=True),
```
Effects specifier should be _after_ `Parameter`. `set(newValue) throws`, not `set throws (newValue)`. It's unfortunate that `test/Parse/effectful_properties.swift` doesn't have cases like it.

rintaro reviewed (commented):
Thanks! Although I understand it's important to always enable syntax parsing, but for this PR, could you hold off enabling `-Xfrontend -verify-syntax-tree` for all tests? Instead, please add test cases to `test/Syntax`. Before enabling `-verify-syntax-tree` everywhere, I'd like to ensure it doesn't change any frontend behavior. For example, it changes the behavior of `Lexer` whether or not it advances the cursor until the end of the trailing trivia after parsing a token. I don't think this causes visible issues, but still.

ahoppen commented:
After some offline discussion, @rintaro and I agreed that it’s to risky enabling `-verify-syntax-tree` for all runtime tests because it changes the Lexer’s and Parser’s behaviors slightly (trivia are lexed into leading and trailing if enabled and delayed function body parsing is disabled) and we would thus not be testing the code path used during “normal” compilation.

Instead, we agreed that
* we use this PR to merge the fixes I found by enabling `-verify-syntax-tree` for `%target-build-swift`
* I run the test suite with `-verify-syntax-tree` enabled locally from time to time
* once we have migrated the parser to use libSyntax, we will always be creating a libSyntax tree so the issue described above will disappear and we can enable `-verify-syntax-tree` for `%target-build-swift`

ahoppen committed:
[libSyntax] Several fixes for libSyntax parsing

Fixe a couple of bugs in libSyntax parsing found by enabling `-verify-syntax-tree` for `%target-build-swift`:
- Fix parsing of the `actor` contextual keyword in actor decls
- Don't build a libSyntax tree when parsing the availability macro
  - The availability macro is not part of the source code and doesn't form a valid Swift file, thus creation of a libSyntax tree is completely pointless and will fail
- Add support for parsing `@_originallyDefinedIn` attributes.
- Add support for parsing `#sourceLocation` in member decl lists
- Add support for effectful properties (throwing/async getters/setters)
- Add support for optional types as the base of a key path (e.g. `\TestOptional2?.something`)
- Allow platform restrictions without a version (e.g. `_iOS13Aligned`)

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
apple/swift-syntax#274

@swift-ci Please smoke test

rintaro reviewed (approved):
LGTM!

ktoso commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -4421,6 +4430,7 @@ Parser::parseDecl(ParseDeclOptions Flags,
 
     if (shouldParseExperimentalConcurrency() &&
         Tok.isContextualKeyword("actor") && peekToken().is(tok::identifier)) {
+      Tok.setKind(tok::contextual_keyword);
```
that's right AFAIK as I confirmed with the team a few weeks ago 👍

ktoso reviewed (approved):
Thanks a lot, seems good to me :)
